includes/payment-methods/class-wc-stripe-upe-payment-gateway.php

```diff
@@ -521,51 +686,192 @@ public function process_upe_redirect_payment( $order_id, $intent_id, $save_payme
+			$charge = end( $intent->charges->data );
+			$this->process_response( $charge, $order );
+		} else {
+			// TODO: Add implementation for setup intents.
 			$this->process_response( $intent, $order );
+		}
+		$this->save_intent_to_order( $order, $intent );
+		$this->set_payment_method_title_for_order( $order, $payment_method_type, $payment_method_details );
+	}
 
-			$this->set_payment_method_title_for_order( $order, $payment_method_type, $payment_method_details );
+	/**
+	 * Converts payment method into object similar to prepared source
+	 * compatible with wc_stripe_payment_metadata and wc_stripe_generate_payment_request filters.
+	 *
+	 * @param object           $payment_method Stripe payment method object response.
+	 * @param WC_Payment_Token $token          WC Payment Token.
+	 *
+	 * @return object
+	 */
+	public function prepare_payment_method( $payment_method, $token ) {
+		return (object) [
+			'token_id'              => $token->get_id(),
+			'customer'              => $payment_method->customer,
+			'source'                => null,
+			'source_object'         => null,
+			'payment_method'        => $payment_method->id,
+			'payment_method_object' => $payment_method,
+		];
+	}
 
-		} catch ( Exception $e ) {
-			WC_Stripe_Logger::log( 'Error: ' . $e->getMessage() );
+	/**
+	 * Save payment method to order.
+	 *
+	 * @param WC_Order $order For to which the source applies.
+	 * @param stdClass $payment_method Stripe Payment Method.
+	 */
+	public function save_payment_method_to_order( $order, $payment_method ) {
+		if ( $payment_method->customer ) {
+			$order->update_meta_data( '_stripe_customer_id', $payment_method->customer );
+		}
+		$order->update_meta_data( '_stripe_payment_method_id', $payment_method->id );
 
-			/* translators: localized exception message */
-			$order->update_status( 'failed', sprintf( __( 'UPE payment failed: %s', 'woocommerce-gateway-stripe' ), $e->getMessage() ) );
+		if ( is_callable( [ $order, 'save' ] ) ) {
+			$order->save();
+		}
+	}
 
-			wc_add_notice( $e->getMessage(), 'error' );
-			wp_safe_redirect( wc_get_checkout_url() );
-			exit;
+	/**
+	 * Retries the payment process once an error occured.
+	 *
+	 * @param object   $intent     The Payment Intent response from the Stripe API.
+	 * @param WC_Order $order      An order that is being paid for.
+	 * @param bool     $can_retry
+	 * @throws WC_Stripe_Exception If the payment is not accepted.
+	 * @return array|void
+	 */
+	public function retry_after_error( $intent, $order, $can_retry ) {
+		if ( ! $can_retry ) {
+			$localized_message = __( 'Sorry, we are unable to process your payment at this time. Please retry later.', 'woocommerce-gateway-stripe' );
+			$order->add_order_note( $localized_message );
+			throw new WC_Stripe_Exception( print_r( $intent, true ), $localized_message ); // phpcs:ignore WordPress.PHP.DevelopmentFunctions.
+		}
+
+		// Don't do anymore retries after this.
+		if ( 5 <= $this->retry_interval ) {
```

Comment:
retry_interval property does not exist. It's in  WC_Gateway_Stripe. Probably it should be moved to WC_Stripe_Payment_Gateway with a default value.

Reply:
Nice one. In ba8ba6ba2799aeaeb9bf9023ce9060000f48ab7a, I moved this to the `WC_Stripe_Payment_Gateway` base class as a protected property with a default value. A couple other classes inherit this and set the value to `1` in their constructors, so I removed those redundant initial assignments: it gets set to `2` in `WC_Stripe_Webhook_Handler` over [here](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/ba8ba6ba2799aeaeb9bf9023ce9060000f48ab7a/includes/class-wc-stripe-webhook-handler.php#L35), so I left that one in.